### DISCUSSION TITLE
useLazyAtom

### DISCUSSION THREAD

**Discussion by AleksandrukTad** [pmndrs/jotai#1787] [Ideas] [answered]
Hey, not sure if this feature is already implement or not but I think its worth looking into.

I have a form
![image](https://user-images.githubusercontent.com/22525952/220879130-00c220e4-5365-4087-8395-191d1d276c8c.png)

- Form container is responsible for fetching
- Form component is handling the logic for stepper, does some additional logic for file upload. Moreover it does validation for Form data. Plus is pulls the useSetAtom to validate, changeSomeFields and clearErrors.
- Steps are parts of the form. They pull atoms to read/write data.

Since **form container** is responsible for fetching I would like to be able to get values from the form **without** creating subscription. My approach is as follows:

In atoms

```
export const getLazyIntegrationAtom = atom(
  null,
  (get, set) => {
    const integration = get(integrationAtom)

    return integration;
  }
) 
```

In component

```
  const getIntegration = useSetAtom(getLazyIntegrationAtom);

  const handleCreateIntegration = async () => {
    try {
      const integration = getIntegration();
      createIntegrationMutation.mutateAsync({
        ...integration,
      })

      push("/app")
    } catch (error) {
      console.error(error)
    }
  }
```

Is this approach alright?

I have another idea how to tackle this is by returning the form fields form the validation and then pass it to handleSuccess like so:

In component:

```
  const handleSubmit = () => { 
    const [fields, errors]= validate({ zodModel: IntegrationRequestBodyDto })

    if (!errors) onSuccess(fields)
  }
```

In container:
```
  const handleCreateIntegration = async (fields) => {
    try {
      const integration = getIntegration();
      createIntegrationMutation.mutateAsync({
        ...fields,
      })

      push("/app")
    } catch (error) {
      console.error(error)
    }
  }
```

**Comment by dai-shi** [comment, marked as answer]
Your approach is valid. (It wasn't supported in v1, but v2 does.)

Alternatively, you can create a `handleCreateIntegrationAtom`.
```js
const handleCreateIntegrationAtom = atom(null, async (get, set) => {
  try {
    const integration = get(integrationAtom)
    createIntegrationMutation.mutateAsync({
      ...fields,
    })
  // ...
})
```
If it needs to read props or other states, you can define the atom in a component using `useMemo`.
Alternatively, `useAtomCallback` util does the same thing.

**Reply by AleksandrukTad** [reply]
And this `handleCreateIntegrationAtom` can be defined inside component?

**Reply by dai-shi** [reply]
If you need, you can define the atom in a component using useMemo.
